2.9 upgrade: exception in Play java async actions that return CompleteableFuture.completedStage

<!--- Copyright (C) from 2022 The Play Framework Contributors <https://github.com/playframework>, 2011-2021 Lightbend Inc. <https://www.lightbend.com> -->

When upgrading to Play 2.9, returning a `CompletableFuture.completedStage()` in Java controller action causes the framework to throw an exception. Changing the code to return a `CompletableFuture.completedFuture()` avoids the exception.

### Play Version

2.9-RC2

### API

Java

### Operating System

Alpine Linux, [eclipse-temurin:11.0.20_8-jdk-alpine](https://github.com/civiform/civiform/blob/main/Dockerfile#L7)

### JDK

Java 11.0.20_8

Scala 2.13.12

### Library Dependencies

Libraries listed here: https://github.com/civiform/civiform/blob/main/server/build.sbt

### Expected Behavior

<!-- Please describe the expected behavior of the issue, starting from the first action. -->

1. upgrade to Play 2.9-RC2
2. make a request to controllers with dependencies that return a `CompletableFuture.completedStage`
3. request succeeds normally

### Actual Behavior

Play Framework throws an exception:

```
Caused by: java.lang.UnsupportedOperationException: null
	at java.base/java.util.concurrent.CompletableFuture$MinimalStage.isDone(CompletableFuture.java:2844)
	at scala.jdk.javaapi.FutureConverters$.asScala(FutureConverters.scala:73)
	at scala.jdk.FutureConverters$CompletionStageOps$.asScala$extension(FutureConverters.scala:41)
	at play.core.j.JavaAction$$anon$1.$anonfun$call$1(JavaAction.scala:127)
	at play.api.mvc.BodyParser$.$anonfun$parseBody$4(Action.scala:241)
	at scala.Option.getOrElse(Option.scala:201)
	at play.api.mvc.BodyParser$.parseBody(Action.scala:239)
	at play.core.j.JavaAction$$anon$1.call(JavaAction.scala:128)
	at play.http.DefaultActionCreator$1.call(DefaultActionCreator.java:31)
	at org.pac4j.play.java.SecureAction.lambda$internalCall$1(SecureAction.java:100)
	at org.pac4j.core.engine.DefaultSecurityLogic.perform(DefaultSecurityLogic.java:141)
	at org.pac4j.play.java.SecureAction.internalCall(SecureAction.java:94)
	at org.pac4j.play.java.SecureAction.call(SecureAction.java:77)
	at play.core.j.JavaAction.$anonfun$apply$8(JavaAction.scala:184)
	at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:687)
	at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:467)
	at play.core.j.ClassLoaderExecutionContext.$anonfun$execute$1(ClassLoaderExecutionContext.scala:64)
	at play.api.libs.streams.Execution$trampoline$.executeScheduled(Execution.scala:104)
	at play.api.libs.streams.Execution$trampoline$.execute(Execution.scala:66)
	at scala.concurrent.impl.Promise$Transformation.submitWithValue(Promise.scala:429)
```

### Reproducible Test Case

The code changes to `QuestionService.java` in this commit fixed the issue: https://github.com/civiform/civiform/pull/5648/commits/46ed43981901cc8c3dc6a6367a83192871e7468e

Previously in [the associated PR](https://github.com/civiform/civiform/pull/5648) we worked around this by making the action synchronous. Changing `completedStage` to `completedFuture` allowed us to change it back to being async.

--- a/core/play-integration-test/src/test/scala/play/it/http/JavaActionSpec.scala
+++ b/core/play-integration-test/src/test/scala/play/it/http/JavaActionSpec.scala
@@ -4,6 +4,9 @@
 
 package play.it.http
 
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.CompletionStage
+
 import akka.util.ByteString
 import play.api.inject.guice.GuiceApplicationBuilder
 import play.api.libs.ws._
@@ -24,7 +27,7 @@ import play.routing.{ Router => JRouter }
 class GuiceJavaActionSpec extends JavaActionSpec {
   override def makeRequest[T](
       method: String,
-      controller: MockController,
+      controller: AbstractMockController,
       configuration: Map[String, AnyRef] = Map.empty,
       body: WSBody = EmptyBody
   )(block: WSResponse => T): T = {
@@ -50,7 +53,7 @@ class BuiltInComponentsJavaActionSpec extends JavaActionSpec {
 
   override def makeRequest[T](
       method: String,
-      controller: MockController,
+      controller: AbstractMockController,
       configuration: Map[String, AnyRef] = Map.empty,
       body: WSBody = EmptyBody
   )(block: (WSResponse) => T): T = {
@@ -87,11 +90,20 @@ class BuiltInComponentsJavaActionSpec extends JavaActionSpec {
 trait JavaActionSpec extends PlaySpecification with WsTestClient {
   def makeRequest[T](
       method: String,
-      controller: MockController,
+      controller: AbstractMockController,
       configuration: Map[String, AnyRef] = Map.empty,
       body: WSBody = EmptyBody
   )(block: WSResponse => T): T
 
+  "action" should {
+    "support CompletableFuture.completedStage as return value" in makeRequest(
+      "GET",
+      new AsyncMockController {
+        override def action(request: Request): CompletionStage[Result] =
+          CompletableFuture.completedStage(Results.ok("good"))
+      },
+    ) { response => response.body[String] must beEqualTo("good") }
+  }
   "POST request" should {
     "with no body should result in hasBody = false" in makeRequest(
       "POST",

--- a/core/play/src/main/scala/play/core/j/JavaAction.scala
+++ b/core/play/src/main/scala/play/core/j/JavaAction.scala
@@ -124,7 +124,7 @@ abstract class JavaAction(val handlerComponents: JavaHandlerComponents)
           .parseBody(
             parser,
             request.asScala(),
-            (r: Request[_]) => invocation(r.asJava).asScala.map(_.asScala())
+            (r: Request[_]) => invocation(r.asJava).toCompletableFuture.asScala.map(_.asScala())
           )(executionContext)
           .map(_.asJava)
           .asJava

